### PR TITLE
RootInlineNode

### PR DESCRIPTION
**Pull request by gohabereg**
PR adds:
- RootInlineNode as single point access to inline fragment tree
- Updates directory structure
- Add tree normalization on input and output
- Inline fragments tree integration tests (done, but all tests require refactoring)

### CODE REVIEW

File: src/entities/inline-fragments/specs/InlineTree.integration.spec.ts [hunk 1]
```diff
@@ -0,0 +1,373 @@
+import { RootInlineNode } from '../RootInlineNode';
+import { TextNode } from '../TextNode';
+import { createInlineToolData, createInlineToolName } from '../FormattingNode';
+
+describe('Inline fragments tree integration', () => {
```
Comment (neSpecc) [inline src/entities/inline-fragments/specs/InlineTree.integration.spec.ts:5]:
```suggestion
describe('Inline fragments tree integration', () => {
```

File: src/entities/inline-fragments/specs/InlineTree.integration.spec.ts [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+      const index = 10;
+      const length = 5;
+
+      tree.removeText(index, index + length);
+
+      expect(tree.getText())
+        .toBe(initialText.slice(0, index) + initialText.slice(index + length));
+    });
+
+    it('should remove text from the end of the tree', () => {
```
Comment (neSpecc) [inline src/entities/inline-fragments/specs/InlineTree.integration.spec.ts:106]:
```suggestion
    it('should remove text from passed index to the end of the tree', () => {
```

File: src/entities/inline-fragments/specs/InlineTree.integration.spec.ts [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    });
+
+    it('should throw an error if index is out of range', () => {
+      const initialText = 'Editor.js is a block-styled editor.';
+      const child = new TextNode({ value: initialText });
+      const tree = new RootInlineNode({ children: [ child ] });
+      const index = 100;
+
+      expect(() => tree.removeText(index))
+        .toThrowError();
```
Comment (neSpecc) [inline src/entities/inline-fragments/specs/InlineTree.integration.spec.ts:125]: could we also check the error text or type here?

File: src/entities/inline-fragments/specs/InlineTree.integration.spec.ts [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+          range: [
+            index + offset,
+            index + length + offset,
+          ],
+        },
+      ]);
+    });
+  });
+
+  it('should pass "ultimate" test', () => {
```
Comment (neSpecc) [inline src/entities/inline-fragments/specs/InlineTree.integration.spec.ts:310]: please add a few words of what is happening and asserting here

File: src/entities/inline-fragments/TextNode/index.ts [hunk 5]
```diff
@@ -14,7 +15,7 @@ export class TextNode implements InlineNode {
   /**
    * Private field representing the text content of the node
```
Comment (ilyamore88) [inline src/entities/inline-fragments/TextNode/index.ts:16]: Fix doc here

File: src/entities/inline-fragments/FormattingNode/index.ts [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+      currentFragment.data = this.data;
+    }
+
+    fragments.unshift(currentFragment);
+
+    return fragments;
+  }
+
+  /**
+   * Splits current node by the specified index
```
[the remark this answers is not in the document]
Comment (gohabereg) [reply src/entities/inline-fragments/FormattingNode/index.ts:90]: split is what this method do — splits node into two ones. And it used in the cases where we need to _split_ code. I understand your concerns, but split for array is not split for a tree node

File: src/entities/inline-fragments/mixins/ParentNode/index.ts [hunk 7]
```diff
@@ -142,5 +154,35 @@ export function ParentNode<C extends { new(...args: any[]): InlineNode }>(constr
...
+     */
+    public normalize(): void {
+      const children = Array.from(this.children);
+
+      if (!children.length) {
+        return;
+      }
+
+      children.reduce((prev, child) => {
+        if (prev && prev.isEqual(child)) {
```
[the remark this answers is not in the document]
Comment (gohabereg) [reply src/entities/inline-fragments/mixins/ParentNode/index.ts:170]: No, if data is not equal we don't need to merge nodes — as we won't be able to persist data for both of them

File: src/entities/inline-fragments/specs/InlineTree.integration.spec.ts (hunk 1 above)
Comment (gohabereg) [reply src/entities/inline-fragments/specs/InlineTree.integration.spec.ts:5]: It seems you forgot your suggestion here

File: src/entities/inline-fragments/specs/InlineTree.integration.spec.ts (hunk 3 above)
Comment (gohabereg) [reply src/entities/inline-fragments/specs/InlineTree.integration.spec.ts:125]: It's not a good idea to check for error text as we might change it later. We can use custom error classes, but that's a whole big task

File: src/entities/inline-fragments/FormattingNode/index.ts [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+   *
+   * @param node - node to merge with
+   */
+  public mergeWith(node: InlineNode): void {
+    if (!this.isEqual(node)) {
+      throw new Error('Can not merge unequal nodes');
+    }
+
+    /**
+     * @todo merge data
```
[the remark this answers is not in the document]
Comment (gohabereg) [reply src/entities/inline-fragments/FormattingNode/index.ts:214]: That should be decided by the tool. There might be cases where data is not equal but still mergable

File: src/entities/inline-fragments/ParentInlineNode/index.ts [hunk 9]
```diff
@@ -1,50 +1,29 @@
...
  */
 @ParentNode
-@ChildNode
-export class FormattingNode implements InlineNode {
-  /**
-   * Private field representing the name of the formatting tool applied to the content
-   */
-  #tool: InlineToolName;
-
+export class ParentInlineNode implements InlineNode {
```
[the remark this answers is not in the document]
Comment (gohabereg) [reply src/entities/inline-fragments/ParentInlineNode/index.ts]: Abstract here breaks the types. Anyway, there is no abstract methods in the class, an you can actually create an instance of ParentInlineNode — it doesn't break any logic

File: src/entities/BlockNode/index.ts [hunk 10]
```diff
@@ -38,7 +38,7 @@ export class BlockNode {
   /**
    * Constructor for BlockNode class.
    *
-   * @param args - TextNode constructor arguments.
+   * @param args - TextInlineNode constructor arguments.
```
Comment (ilyamore88) [inline src/entities/BlockNode/index.ts:41]:
```suggestion
   * @param args - BlockNode constructor arguments.
```

File: src/entities/inline-fragments/FormattingInlineNode/types/InlineToolName.ts
Comment (ilyamore88) [inline src/entities/inline-fragments/FormattingInlineNode/types/InlineToolName.ts:1]: Update docs in this file

**Review by ilyamore88 — APPROVED** [review]

File: src/entities/inline-fragments/specs/InlineTree.integration.spec.ts (hunk 3 above)
Comment (neSpecc) [reply src/entities/inline-fragments/specs/InlineTree.integration.spec.ts:125]: If we change the error text, we can also change it in tests. For now any error thrown during execution of `removeText()` will lead false-positive test assertion

**Review by neSpecc — APPROVED** [review]

**Pull request merged**